### DISCUSSION TITLE
Jamulus server on Raspberry Pi Cluster

### DISCUSSION THREAD

**Discussion by MERE55** [jamulussoftware/jamulus#1637] [Technical Support & Help] [answered]
As I am planning to create a Raspberry Pi cluster, e.g. with 4x Raspberry Pi 4B I thought of running a Jamuus server on that envoironment.
Would it be possible and improve the power of the Jamulus server / increase the number of possible attached Jamulus clients when using such a Raspberry Pi cluster with a Kubernetes architecture ?

**Comment by hoffie** [comment, marked as answer]
Jamulus should be able to run in containers and so it should be able to schedule them using Kubernetes. However, there is no mechanism for linking multiple Jamulus instances to form a single cluster.

So if your goal is to simplify individual Jamulus server provisioning, this may work.
If your goal is to distribute the load of a single Jamulus instance (where everyone can hear each other), then it will not work.

**Reply by MERE55** [reply]
Well, the goal would have been to enable more Jamulus clients to take part at a choir rehearsal than with a single Raspberry Pi.
But when I undderstood you correct that seems not to work, right ?

**Reply by hoffie** [reply]
I just ran a quick test with my Raspberry Pi 3 Model B Plus Rev 1.3 and it was able to handle 8 concurrent Mono channels with Normal quality. After that, ping started spiking.
No tuning though, also running other stuff. I'm getting the same results with and without `-T`.

**Reply by dcorson-ticino-com** [reply]
How many cores does a Pi3 have?
At least on the Odroid you see immediately with top that using -T the work load is spread out across the 4 fast cores in comparison to without -T.

**Reply by hoffie** [reply]
> How many cores does a Pi3 have?

4.

> At least on the Odroid you see immediately with top that using -T the work load is spread out across the 4 fast cores in comparison to without -T.

Hadn't checked, but may be the bottleneck is not opus but something else (network interrupts or something).

**Comment by dcorson-ticino-com** [comment]
The problem will be that all the server client inputs are mixed and then sent back to all the clients.  You can not give each processor a subset of the clients.
The present multi-threading in Jamulus uses the fact that there are several steps during the audio processing that are done sequentially giving each step to a different core.  But my hunch is that the time lost in communications between the members of the cluster will be too great to allow this to work across processors that don't have shared memory.

**Reply by MERE55** [reply]
Thx for reply - that seems that it would not be possible to enable more participants to join a rehearsal than with only one Raspberry Pi.
How many singers do you think could join a single Raspberry Pi Jamulus server ?

**Reply by dcorson-ticino-com** [reply]
I think that you would have to test that as it depends also on your Internet connection, but we have heard reports of between 10 and 15 clients on a RaspPi server.
I run a server on an Odroid-N2, like a Pi it uses an ARM processor, but is about twice the speed of a Pi4.  My tests suggest that I could do about 35 clients, but I have limited to 25 to not have any surprises.  There have been successful rehearsals with 25 clients so I think those numbers for the Pi are reasonable.

**Reply by MERE55** [reply]
So wih this you say that it is more a matter of speed / performance of a single processor ...
Then what would you mean to a ASUS Tinker Board ?
Would that be even more performant than your Odroid-N2 server and could provide capacity for more Jamulus `clients ?

**Reply by dcorson-ticino-com** [reply]
Hi Mere,
I am not familiar with that board, but if it is faster, why not ?
I chose the Odroid-N2 at the time because it does not need any cooling fan with its big heat sink.  A big advantage in my book.
Be sure to run with multi-threading turned on "-T".
Have fun!
